keps/sig-node/0029-2018-09-19-node-fencing.md

```diff
@@ -0,0 +1,223 @@
+* Clayton Colman [Pod Safety](https://github.com/kubernetes/community/blob/16f88595883a7461010b6708fb0e0bf1b046cf33/contributors/design-proposals/pod-safety.md) October 2016
+
+
+### Goals
+
+* Specify where implementations should look in order to determine if a node requires fencing
+* Specify how implementations should indicate a node needs to be fenced
+* Specify how implementations should indicate a node has been sucessfully fenced
+* Specify how nodes should return to a "normal" state
+* Specify the changes required to Kubernetes that enable automated recovery in the presense of worker failures
+* Provide a sample implementation based on the [Machine](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/machine-api-proposal.md) API 
+
+### Non-Goals
+
+It is explicitly _not_ the goal of this work to:
+
+* Have fencing become a requirement for all Kubernetes deployments
+* Have the community converge on a particular fencing implementation
+
+## Proposal
+
+Three additional NodeConditions and taints:
+
+* a `node.alpha.kubernetes.io/fencingtriaged` taint corresponding to NodeCondition `FencingTriaged` being
+   `True`
+* a `node.alpha.kubernetes.io/fencingrequired` taint corresponding to NodeCondition `FencingRequired` being
+   `True`
+* a `node.alpha.kubernetes.io/fencingcomplete` taint corresponding to NodeCondition `FencingComplete` being
+   `True`
+
+This allows administrators and other system components to differentiate between:
+
+* Nodes expected to be operational () 
+* Nodes in an error state that have been triaged but do not require fencing (yet) (`FencingTriaged` = `True`, `FencingRequired` = `False` )
+* Nodes in an error state that have been triaged (`FencingTriaged` = `True`, `FencingRequired` = `True` )
+* Nodes in an error state that are being recovered (`FencingTriaged` = `True`, `FencingRequired` = `True`, `FencingComplete` = `False` )
+* Nodes with errors that have entered a safe state (`FencingTriaged` = `True`, `FencingRequired` = `True`, `FencingComplete` = `True` )
+
+Setting and unsetting the NodeConditions form the primary contract between
+implementations and the rest of the system.
+
+1. Implementations should use NodeConditions in order to determine if a node requires fencing.
+
+   If additional information is required, it should be made available by
+   augmenting the existing the list of NodeConditions using something like
+   [Node Problem Detector](https://github.com/kubernetes/node-problem-detector/)
+
+   Including error condition detection in fencing implementations should be
+   avoided.
+
+1. Implementations should indicate a Node needs to be fenced with the `NeedsFencing` NodeCondition.
+
+   Admins may also set NodeCondition `FencingRequired` to `True` to manually
+   trigger fencing.
+
+   Once fencing has been initiated, the NodeCondition `FencingComplete` should
```

Comment:
"should" is a bad word.

Who will set this?

Reply:
I would expect this to be handled by the fencing implementation.
I will replace s/should/must/